### PR TITLE
GCS 503 retry [BA-5881]

### PR DESCRIPTION
**Pull request by aednichols**

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/io/IoActor.scala [hunk 1]
```diff
@@ -179,6 +179,11 @@ object IoActor {
     case _ => false
   }
 
+  def is5xxServerError(failure: Throwable): Boolean = {
+    val serverErrorPattern = "^.*5\\d{2} \\w+(.*)?"
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/io/IoActor.scala]: This seems like a fairly loose regex, a 5 and two digits followed by a space and at least one word char, preceded by anything and followed by anything. No concern about false positives?

Comment (aednichols) [reply engine/src/main/scala/cromwell/engine/io/IoActor.scala]: The exact error class we're targeting (from @dinvlad) is
```
Bad output 'get_segs.count': Failed to read_int("count.txt") (reason 1 of 1): [Attempted 1 time(s)] - IOException: Could not read from gs://broad-epi-cromwell/workflows/ChipSeq/a4a8808f-32f3-4f3e-84f2-04fd7861c581/call-get_segs/shard-0/count.txt: 503 Service Unavailable Backend Error
```

I suppose we could be conservative by making it super specific and matching only on `503 Service Unavailable Backend Error`

**Comment by aednichols** [comment]
>Does "5xx HTTP Status Code" literally appear in real life instead of the actual 500-511 error codes?
No but I think the idea is to make sure the regex that checks for "5" matches "500" but not "5 apples"

**Comment by aednichols** [comment]
Updated to be way more specific in targeting Denis-reported errors. I've found this area can get difficult to reason about when we throw in ranges and very broad regexes, so I chose to be incredibly specific at the cost of more code & slightly less functionality (but Denis's functionality is there).

File: engine/src/main/scala/cromwell/engine/io/IoActor.scala [hunk 2]
```diff
@@ -179,6 +179,16 @@ object IoActor {
     case _ => false
   }
 
+  def isGCS500(failure: Throwable): Boolean = {
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/io/IoActor.scala]: the capitalization...

File: engine/src/main/scala/cromwell/engine/io/IoActor.scala [hunk 3]
```diff
@@ -191,26 +201,27 @@ object IoActor {
...
       isRetryable(gcs.getCause) ||
-      AdditionalRetryableHttpCodes.contains(gcs.getCode) || 
+      AdditionalRetryableHttpCodes.contains(gcs.getCode) ||
       AdditionalRetryableErrorMessages.exists(gcs.getMessage.toLowerCase.contains)
     case _: SSLException => true
     case _: BatchFailedException => true
     case _: SocketException => true
     case _: SocketTimeoutException => true
     case ioE: IOException if Option(ioE.getMessage).exists(_.contains("Error getting access token for service account")) => true
+    case ioE: IOException => isGCS500(ioE) || isGCS503(ioE)
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/io/IoActor.scala]: These two feel like they might be better of as either "three separate cases" or "entirely decided within one case"... but YMMV

Comment (dinvlad) [reply engine/src/main/scala/cromwell/engine/io/IoActor.scala]: We've played around a bunch with that, and it seems to only lead to problems..

**Comment by dinvlad** [comment]
Looks like we now have `504 Gateway Timeout` errors too! Should we perhaps add a new case for that?

**Pull request merged**